synapse/handlers/message.py

```diff
@@ -856,8 +878,10 @@ async def create_and_send_nonmember_event(
                 requester,
                 event_dict,
                 txn_id=txn_id,
+                allow_no_prev_events=allow_no_prev_events,
```

Comment:
was the previous omission of this a bug?

Reply:
(could we get rid of `allow_no_prev_events` and just allow there to be no prev events when `state_event_ids` is given?)

Reply:
> was the previous omission of this a bug?

Yes and seems to be why this PR fixes @tulir's issue, https://github.com/matrix-org/synapse/pull/12083#discussion_r816280776

> (could we get rid of `allow_no_prev_events` and just allow there to be no prev events when `state_event_ids` is given?)

The assertion nature of `allow_no_prev_events` disallowing the case for the majority of normal event sending cases is nice.

Also in our case, we only want the first event to have no `prev_event_ids`. The rest in the chain using `state_event_ids` should have `prev_event_ids` pointing at the event before it in the chain.

`state_event_ids` is really just used to derive `auth_event_ids`. Although I do see how they're related, `auth_event_ids` explicitly set by `state_event_ids` or derived from `prev_event_ids` in normal cases.

Reply:
For context: we see a fair few "can't create event with no prev events" errors already, so having an `allow_no_prev_events` flag should hopefully make it easier to reason about what is going on